file-updates/test.sh

```diff
@@ -0,0 +1,159 @@
+    kubectl delete releaseplanadmission "$RELEASE_PLAN_ADMISSION_NAME" "$MANAGED_KUBECONFIG"
+    kubectl delete releasestrategy "$RELEASE_STRATEGY_NAME" "$MANAGED_KUBECONFIG"
+
+    if kubectl get application "$APPLICATION_NAME"  "$DEV_KUBECONFIG" &> /dev/null; then
+        echo "Application $APPLICATION_NAME exists. Deleting..."
+        kubectl delete application "$APPLICATION_NAME" "$DEV_KUBECONFIG"
+    else
+        echo "Application $APPLICATION_NAME does not exist."
+    fi
+}
+
+# Function to watch Build or Release PipelineRun and wait till succeeds.
+function wait_for_pr_to_complete() {
+    local kube_config
+    local type=$1
+    local start_time=$(date +%s)
+
+    if [ "$type" = "release" ]; then
+        kube_config="$MANAGED_KUBECONFIG"
+        crd_labels="appstudio.openshift.io/application=$APPLICATION_NAME,pipelines.appstudio.openshift.io/type=$type"
+    else
+        kube_config="$DEV_KUBECONFIG"
+        crd_labels="appstudio.openshift.io/application=$APPLICATION_NAME,pipelines.appstudio.openshift.io/type=$type,appstudio.openshift.io/component=$COMPONENT_NAME"
+    fi
+
+    while true; do
+        crd_json=$(kubectl get PipelineRun -l "$crd_labels" "$kube_config" -o=json)
+
+        reason=$(echo "$crd_json" | jq -r '.items[0].status.conditions[0].reason')
+        status=$(echo "$crd_json" | jq -r '.items[0].status.conditions[0].status')
+        type=$(echo "$crd_json" | jq -r '.items[0].status.conditions[0].type')
+        name=$(echo "$crd_json" | jq -r '.items[0].metadata.name')
+        namespace=$(echo "$crd_json" | jq -r '.items[0].metadata.namespace')
+
+        if [ "$status" = "False" ] || [ "$type" = "Failed" ]; then
+            echo "PipelineRun $name failed."
+            return 1
+        fi
+
+        if [ "$status" = "True" ] && [ "$type" = "Succeeded" ]; then
+            echo "PipelineRun $name succeeded."
+            return 0
+        else
+            current_time=$(date +%s)
+            elapsed_time=$((current_time - start_time))
+
+            if [ "$elapsed_time" -ge "$TIMEOUT_SECONDS" ] ; then
+                echo "Timeout: PipelineRun $name in namespace $namespace did not succeeded within $TIMEOUT_SECONDS seconds."
+                return 1
+            fi
+            echo "Waiting for PipelineRun $name in namespace $namespace to succeed."
+            sleep 5
+        fi
+    done
+}
+
+OPTIONS=$(getopt -l "skip-cleanup,help" -o "sc,h" -a -- "$@")
+eval set -- "$OPTIONS"
+while true; do
+    case "$1" in
+        -sc|--skip-cleanup)
+            CLEANUP="true"
+            ;;
+        -h|--help)
+            print_help
+            exit
+            ;;
+        --)
+            shift
+            break
+            ;;
+    esac
+    shift
+done
+
+if [ "${CLEANUP}" != "true" ]; then
+  trap teardown EXIT
+else
+  echo "Warning: cleanup at end is disabled, cleanup before setup"
```

Comment:
`cleanup before setup` sounds like a command - something the user should do. I would suggest `cleaning up before setup`

But I'm wondering if the logic is good here. How is cleaning up before the test related to cleaning up after the test? Consider that you want to debug after the test, so you want to disable cleanup at the end. But that doesn't say anything about whether you want to clean up before the test, no?

So maybe by default we should not do anything before and do cleanup after. And have two cli arguments to override it? `--cleanup-before` and `--skip-cleanup-after`? Or maybe the cleanup before should just happen always (and if there's nothing to cleanup, it will just not do anything)?

Reply:
I am thinking cleanup should probably happen always. I cannot think of a case where we want to start test without knowing that the environment is clean. I'll make that change